docs/designs/upgrade-design/upgrade-design-doc.md

```diff
@@ -0,0 +1,653 @@
+**Responsibility**: Execute cluster API migration and validate application compatibility
+**Scope**: Application-specific testing and validation
+**Method**: Update API version in deployment files and execute migration with comprehensive testing
+
+**What gets executed**:
+- Specific DocumentDB cluster API (v1 → v2) migration
+- Application compatibility testing with v2 API features
+- Client application integration validation
+- Data integrity verification during migration
+- Performance testing and optimization
+
+**Database Developer Control**:
+- Executes migration based on Database Admin's coordination plan
+- Gradual rollout across development, staging, production environments
+- Application-specific testing and validation with v2 features
+- Rollback capability per cluster (v2 → v1 downgrade) if application issues occur
+- Test coverage across environments to ensure application compatibility
+
+**Benefits of Three-Tier Responsibility Model**:
+- **Clear Role Separation**: Each team focuses on their domain expertise
+- **Coordinated Migration**: Database Admins coordinate while Database Developers execute
+- **Gradual API Migration**: Database Developers control application-specific migration timeline
+- **Risk Mitigation**: Test new API versions in dev/staging before production migration
+- **Backward Compatibility**: Multiple cluster API versions supported simultaneously
+- **Controlled Deprecation**: API versions deprecated gradually over multiple operator releases
+
+**Cross-Team Coordination**: Database Admins coordinate migration planning with Database Developers for execution, while Kubernetes Admins provide the infrastructure foundation.
+
+**API Version Examples**: See [commands.md](./commands.md) for detailed workflow commands and API version migration examples.
+
+## Component-Specific Upgrade Considerations
+
+While all components upgrade together, each has specific characteristics:
+
+#### Gateway Image Upgrade (API Version Dependent)
+- **State**: **Stateless** - Gateway containers have no persistent state
+- **Impact**: Medium (rolling restart of pods when API version changes)
+- **Risk**: Low - No data loss risk, only temporary connection disruption
+- **Risk Mitigation**: Multiple standby instances with local HA ensure zero-downtime rolling restart; Gateway and PostgreSQL containers run in same pod, sharing HA benefits
+- **Version Behavior**: Gateway features may differ between cluster API v1 and v2
+
+#### PostgreSQL Database Upgrade (API Version Dependent)  
+- **State**: **Stateful** - PostgreSQL contains persistent application data
+- **Impact**: Variable (depends on API version differences)
+- **Risk**: Variable - Data migration only required for breaking schema changes
+- **Risk Mitigation**: CNPG managed HA with supervised rolling updates for zero-downtime upgrades
+- **HA Strategy**: 3-instance clusters (1 primary + 2 standby servers) with CNPG-controlled failover sequence
```

Comment:
why would we need two standby? Cx might only have two or even on AZ in their DC?

Reply:
This is just an example with primary and standby settings. Most common examples use one primary + two standby combinations.